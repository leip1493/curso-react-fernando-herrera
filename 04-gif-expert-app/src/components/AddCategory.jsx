import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState("");

  const handleInputChange = ({ target: { value } }) => {
    setInputValue(value);    
  };

  const handleSubmit = e => {
    e.preventDefault();

    if (inputValue.trim().length > 2) {
      setCategories(categories => [inputValue, ...categories]);
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input type="text" onChange={handleInputChange} value={inputValue} />
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
};
