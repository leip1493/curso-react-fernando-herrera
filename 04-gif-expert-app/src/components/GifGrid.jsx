import React, { useEffect, useState } from "react";
import { GifGridItem } from "./GifGridItem";
import { getGifs } from "../helpers/getGifs";
import { useFetchGifs } from "../hooks/useFetchGifs";

export const GifGrid = ({ category }) => {
  const { gifs, loading } = useFetchGifs(category);

  return (
    <>
      <h3 className="animate__animated animate__fadeIn">{category} </h3>
      {loading && <p className="animate__animated animate__flash">Loading</p>}
      <div className="card-grid">
        {gifs.map(gif => (
          <GifGridItem key={gif.id} {...gif} />
        ))}
      </div>
    </>
  );
};
