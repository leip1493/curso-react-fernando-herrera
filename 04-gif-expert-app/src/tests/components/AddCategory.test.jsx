import React from "react";
import { AddCategory } from "../../components/AddCategory";
import { shallow } from "enzyme";

describe("<AddCategory />", () => {
  const setCategories = jest.fn();
  let wrapper = shallow(<AddCategory setCategories={setCategories} />);

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={setCategories} />);
  });

  test("Debe renderizar exitosamente ", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe cambiar el texto del input", () => {
    const input = wrapper.find("input");

    const value = "Hola mundo";

    input.simulate("change", { target: { value } });

    expect(
      wrapper
        .find("p")
        .text()
        .trim()
    ).toBe(value);
  });

  test("No debe realizar el cambio de estado onSubmit si no hay inputValue", () => {
    const form = wrapper.find("form");
    form.simulate("submit", { preventDefault: () => {} });

    expect(setCategories).not.toHaveBeenCalled();
  });

  test("Debe llamar setCategories y limpiar la caja de texto", () => {
    const value = "Lorem ipsum";

    const input = wrapper.find("input");
    input.simulate("change", { target: { value } });

    const form = wrapper.find("form");
    form.simulate("submit", { preventDefault: () => {} });

    expect(input.prop('value')).toBe("");
    expect(setCategories).toHaveBeenCalledTimes(1);
    // Garantiza que sea llamado con una funcion como argumento
    expect(setCategories).toHaveBeenCalledWith( expect.any(Function));


  });
});
