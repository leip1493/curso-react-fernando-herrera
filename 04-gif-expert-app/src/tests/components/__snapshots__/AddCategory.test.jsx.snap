// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCategory /> Debe renderizar exitosamente  1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
